plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEnginePublishOperation.java

```diff
@@ -0,0 +1,270 @@
+    boolean isMoving = false;
+    // check older publish
+    String oldUri = (String) mapping.get(childModule.getId());
+    String jarUri = path.toOSString();
+    if (oldUri != null && jarUri != null) {
+      isMoving = !oldUri.equals(jarUri);
+    }
+    // setup target
+    IPath jarPath = (IPath) path.clone();
+    IPath deployPath = jarPath.removeLastSegments(1);
+    // remove if requested or if previously published and are now serving
+    // without publishing
+    if (isMoving || kind == IServer.PUBLISH_CLEAN || deltaKind == ServerBehaviourDelegate.REMOVED) {
+      if (oldUri != null) {
+        File file = new File(oldUri);
+        if (file.exists()) {
+          file.delete();
+        }
+      }
+      mapping.remove(childModule.getId());
+      if (deltaKind == ServerBehaviourDelegate.REMOVED) {
+        return;
+      }
+    }
+    // check for changes
+    if (!isMoving && kind != IServer.PUBLISH_CLEAN && kind != IServer.PUBLISH_FULL) {
+      IModuleResourceDelta[] delta = server.getPublishedResourceDelta(modules);
+      if (delta == null || delta.length == 0) {
+        return;
+      }
+    }
+    // ensure target directory
+    if (!deployPath.toFile().exists()) {
+      deployPath.toFile().mkdirs();
+    }
+    // do publish
+    IModuleResource[] resources = server.getResources(modules);
+    IStatus[] publishStatus = helper.publishToPath(resources, jarPath, monitor);
+    statusList.addAll(Arrays.asList(publishStatus));
+    // store into mapping
+    mapping.put(childModule.getId(), jarUri);
+  }
+
+  /**
+   * Publish modules as directory.
+   */
+  private void publishDirectory(IPath path, List<IStatus> statusList, IProgressMonitor monitor) throws CoreException {
+    // delete if needed
+    if (kind == IServer.PUBLISH_CLEAN || deltaKind == ServerBehaviourDelegate.REMOVED) {
+      File file = path.toFile();
+      if (file.exists()) {
+        IStatus[] status = PublishHelper.deleteDirectory(file, monitor);
+        statusList.addAll(Arrays.asList(status));
+      }
+      // request for remove
+      if (deltaKind == ServerBehaviourDelegate.REMOVED) {
+        return;
+      }
+    }
+    // republish or publish fully
+    if (kind == IServer.PUBLISH_CLEAN || kind == IServer.PUBLISH_FULL) {
+      IModuleResource[] resources = server.getResources(modules);
+      IStatus[] publishStatus = helper.publishFull(resources, path, monitor);
+      statusList.addAll(Arrays.asList(publishStatus));
+      return;
+    }
+    // publish changes only
+    IModuleResourceDelta[] deltas = server.getPublishedResourceDelta(modules);
+    for (IModuleResourceDelta delta : deltas) {
+      IStatus[] publishStatus = helper.publishDelta(delta, path, monitor);
+      statusList.addAll(Arrays.asList(publishStatus));
+    }
+  }
+
+  /**
+   * Publish child modules as directory if not binary.
+   */
+  private void publishDir(IPath path, Properties mapping, List<IStatus> statusList, IProgressMonitor monitor,
```

Comment:
Is the monitor actually used here? I don't see it, but if it is, should probably convert to a submonitor.